### ISSUE TITLE
Multihead Attention with mask producing float32 spontaneously, somehow compile cache related

### ISSUE THREAD

**Issue by IlanCosman** [pytorch/pytorch#144117]
### 🐛 Describe the bug

Here is a minimal reproducer. It compiles but produces a warning about float32 usage.
```
UserWarning: TensorFloat32 tensor cores for float32 matrix multiplication available but not enabled. Consider setting `torch.set_float32_matmul_precision('high')` for better performance.
```

```python
import torch
import torch._inductor.config
from torch import Tensor, nn

torch.set_default_dtype(torch.bfloat16)
# Set this to True to produce the bug just the first time you run the script
# Set this to False to always reproduce the bug
torch._inductor.config.fx_graph_cache = True

batch = 2
seq = 3
features = 4

data = torch.randn(batch, seq, features).cuda()
mask = torch.randn(batch, seq).cuda() < 0
mask_per_seq = mask.unsqueeze(1).expand(-1, seq, -1)  # (batch, seq, seq)


@torch.compile
class MyModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.mha = nn.MultiheadAttention(
            embed_dim=features, num_heads=1, batch_first=True
        )

    def forward(self, x: Tensor, attn_mask: Tensor) -> Tensor:
        return self.mha(x, x, x, attn_mask=attn_mask, need_weights=False)[0]


model = MyModule().cuda()
print(model(data, mask_per_seq))
```

This warning shouldn't appear since everything ought to be in bfloat16.

Passing an `attn_mask` is what seems to cause the problem. `need_weights` and `batch_first` etc. don't matter.

Finally, there is some sort of caching component, where the first run will produce the warning, and subsequent runs will not. If one turns off the cache, either from the command line or in the code, the warning will appear every time.

### Versions

Arch Linux
Python 3.12.7
Pytorch 2.5.1+cu124 (also reproduced on nightly 2.6.0.dev20241231+cu126)

cc @chauhang @penguinwu

**Comment by IlanCosman** [comment]
Found the cause:

From the SDPA doc:

> For math backend, all intermediates are kept in torch.float if inputs are in torch.half or torch.bfloat16.
